### PR TITLE
Get count param

### PR DESCRIPTION
**Pull request by tazwar22**
Addressing #50 

Current Implementation:
- Simply returns the first "count" (1,2,3... N) articles (specified in route parameter)
- Not robust to count > collection size
- 1 route Handler for both queried/non-queried versions of GET Article 

Open to suggestions for better design.

### CODE REVIEW

File: backend/routes.go [hunk 1]
```diff
@@ -25,7 +26,9 @@ func (s *server) handleRoutes() error {
 
 	router.HandleFunc("/users/", s.handleAddUser()).Methods("POST")
 	router.HandleFunc("/users/{id}", s.handleGetUser()).Methods("GET")
-	router.HandleFunc("/articles", s.handleGetArticles()).Methods("GET")
+	//Adding route param...
+	router.HandleFunc("/articles/", s.handleGetArticles()).Methods("GET")
+	router.HandleFunc("/articles/{count}", s.handleGetArticles()).Methods("GET")
```
Comment (tarikeshaq) [inline backend/routes.go]: Sweet! I wonder if we can use optional query parameters instead of a path.. i.e something like this https://stackoverflow.com/questions/43379942/how-to-have-an-optional-query-in-get-request-using-gorilla-mux

I know I said to use `vars` in the issue, but this seems a little neater since they are optional (and otherwise we would return all articles if it's not provided)

That way, you can have something like this in the handler,
```
v := r.URL.Query()
count := v.Get("count")
// ... We can add the strconv
```

And we would avoid setup an extra `handleFunc` up here

File: backend/routes.go [hunk 2]
```diff
@@ -93,18 +98,34 @@ func (s *server) handleGetArticles() http.HandlerFunc {
...
+		//ASSUMPTION: count <= SIZE(RESULTS); will fix this later
+		numArticles, err := strconv.Atoi(mux.Vars(req)["count"])
+		if numArticles == 0 { //NO COUNT SPECIFIED
+			fmt.Println("Getting ALL Articles...")
+			numArticles = len(results)
+		}else{
+			fmt.Printf("Limiting number of articles to  %d \n", numArticles)
+		}
+
+		for _, a := range results[:numArticles] {
```
Comment (tarikeshaq) [inline backend/routes.go]: We spoke about it in slack - let's try to do the counting in the database instead of loading all the values into memory first!

File: backend/routes.go [hunk 3]
```diff
@@ -93,18 +98,34 @@ func (s *server) handleGetArticles() http.HandlerFunc {
 
 		var articles = make([]ArticleJson, 0)
 		var article ArticleJson
-		for _, a := range results {
+
+		
+		//Get number of articles necessary
+		//ASSUMPTION: count <= SIZE(RESULTS); will fix this later
+		numArticles, err := strconv.Atoi(mux.Vars(req)["count"])
```
Comment (tarikeshaq) [inline backend/routes.go]: The linter is complaining because we aren't checking the error value, you probably already have that in mind, just wanted to clarify what the liner is complaining about 😄

**Review by tarikeshaq — COMMENTED** [review]
Good work @tazwar22 ! Dropped a couple of comments, if you'd like me to comment on anything specific, let me know

**Comment by tazwar22** [comment]
@tarikeshaq 
Updated implementation:
1.  Followed suggestion for using **Optional** Query parameters; can extend this logic to other routes in the future as need be
2. ` mongo.db` Article Retrieval function only collects the first N-articles
3. Deals with count > N appropriately

There's probably a better way to Iterate over a limited subset of the Collection, compared to the very simple way in which I implemented it, so there's significant room for refactoring the function! Open to an alternative suggestion.

File: backend/db/db.go [hunk 4]
```diff
@@ -63,6 +63,8 @@ func (c *Database) CreateNewArticle(url string, title string, source string) (st
 }
 
 // GetAllArticles gets all articles from db, formatted to Article type
-func (c *Database) GetAllArticles() ([]Article, error) {
-	return c.FindAllArticles()
+// numArticles: number of Articles to return from Collection,
+// NOTE:  If -1 is passed in, we return ALL Articles
+func (c *Database) GetAllArticles(numArticles int) ([]Article, error) {
```
Comment (tarikeshaq) [inline backend/db/db.go:68]: Hmm can we define a seperate function for this?

The name `GetAllArticles` suggests we are getting all articles. Lets define a separate function that takes in a parameter. We can do some refactoring in `mongo.go` to help us with the common code

This way we don't have to carry a special behavior for the value of `-1`

File: backend/db/mongo.go [hunk 5]
```diff
@@ -41,8 +40,11 @@ func (c *Database) insertArticle(article Article) (string, error) {
 // 	return User{}, nil
 // }
 
-// fetch all Article from mongo
-func (c *Database) FindAllArticles() ([]Article, error) {
+
+
+// fetch ALL Article from mongo if numArticles == -1 (FLAG)
```
Comment (tarikeshaq) [inline backend/db/mongo.go:45]: I appreciate the comments! Let's clean them up (we don't need the unnecessary capitalization, and it can be in plain English) once we've refactored :)

Hopefully, we can also later document all our exported functions this way for go doc :)

File: backend/db/mongo.go [hunk 6]
```diff
@@ -57,17 +59,28 @@ func (c *Database) FindAllArticles() ([]Article, error) {
 	}
 	var articles []Article
 
+	numObtained := 0
 	for cursor.Next(context.TODO()) {
```
Comment (tarikeshaq) [inline backend/db/mongo.go:63]: I would prefer if we have a loop where we put numObtained in the condition, saves us from the extra if statement. And assuming we have this in a separate function to get rid of the -1 special handling all will be good
```go
for cursor.Next(context.TODO()) && numObtained < numArticles {
.
.
}

```

File: backend/db/mongo.go [hunk 7]
```diff
@@ -57,17 +59,28 @@ func (c *Database) FindAllArticles() ([]Article, error) {
...
+		//Return early if we have found the necessary number of articles
+		if numArticles != -1 && numObtained >= numArticles {
+			return articles, nil
+		}
 	}
 	return articles, nil
-
 }
 
+
```
Comment (tarikeshaq) [inline backend/db/mongo.go:79]: Lets get rid of this extra spacing. One empty line is ok

File: backend/routes.go [hunk 8]
```diff
@@ -25,7 +26,9 @@ func (s *server) handleRoutes() error {
 
 	router.HandleFunc("/users/", s.handleAddUser()).Methods("POST")
 	router.HandleFunc("/users/{id}", s.handleGetUser()).Methods("GET")
-	router.HandleFunc("/articles", s.handleGetArticles()).Methods("GET")
+	//Adding route param...
+	router.HandleFunc("/articles/", s.handleGetArticles()).Methods("GET")
+	// router.HandleFunc("/articles/{count}", s.handleGetArticles()).Methods("GET")
```
Comment (tarikeshaq) [inline backend/routes.go:31]: We can get rid of those comments
```suggestion
```

File: backend/routes.go [hunk 9]
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
 //GET: endpoint for all articles
 func (s *server) handleGetArticles() http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
+
```
Comment (tarikeshaq) [inline backend/routes.go:84]: Empty lines are unnecessary at the top. At most, we should have one empty line between every logically independent piece of code. Smaller functions are easier to look at :)

File: backend/routes.go [hunk 10]
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
...
+		v := req.URL.Query()
+		//DEFAULT FLAG = -1
+		var numArticles = -1
 
+		//No count value? RETURN ALL
+		if v.Get("count") != "" {
+			numArticles, _ = strconv.Atoi(v.Get("count"))
+		}
+
+		//IF
```
Comment (tarikeshaq) [inline backend/routes.go:101]:
```suggestion
```

File: backend/routes.go [hunk 11]
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
...
 		}
 
-		results, err := s.db.GetAllArticles()
+		v := req.URL.Query()
+		//DEFAULT FLAG = -1
+		var numArticles = -1
 
+		//No count value? RETURN ALL
+		if v.Get("count") != "" {
+			numArticles, _ = strconv.Atoi(v.Get("count"))
```
Comment (tarikeshaq) [inline backend/routes.go:98]: Lets not discard the error. It could be a valid issue with the caller sending an invalid count.

We should return an error to the caller in this case with `http.StatusBadRequest`

File: backend/routes.go [hunk 12]
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
...
+		//IF 
+		results, err := s.db.GetAllArticles(numArticles)
 		if err != nil {
 			s.respond(w, req, makeErrorResponse(err), 500)
 			return
 		}
 
 		var articles = make([]ArticleJson, 0)
 		var article ArticleJson
+
```
Comment (tarikeshaq) [inline backend/routes.go:110]:
```suggestion
```

File: backend/routes.go [hunk 13]
```diff
@@ -101,10 +116,25 @@ func (s *server) handleGetArticles() http.HandlerFunc {
 			}
 			articles = append(articles, article)
 		}
+
 		s.respond(w, req, articles, http.StatusOK)
 	}
 }
 
+
```
Comment (tarikeshaq) [inline backend/routes.go:124]: Ditto on too many unnecessary empty lines

File: backend/db/db.go [hunk 14]
```diff
@@ -63,6 +63,8 @@ func (c *Database) CreateNewArticle(url string, title string, source string) (st
...
+		return c.FindAllArticles(numArticles)
```
Comment (tarikeshaq) [inline backend/db/db.go:69]: Too much indentation here

**Review by tarikeshaq — COMMENTED** [review]
Sweet work @tazwar22 I really like where this is heading 🔥 💯 

I dropped a couple of comments, but keep going!

A few of them can be solved by running `go fmt` before committing your code, `go fmt` will format all the go code in the directory it is run so that it is by go standards! (getting rid of extra white spaces etc)

In your case, since you have changes on both the `backend` directory and the `db` directory, run it in both - it will automagically format the code!
